ema_workbench/em_framework/evaluators.py

```diff
@@ -415,6 +416,81 @@ def evaluate_experiments(self, scenarios, policies, callback, combine="factorial
+experiment_runner = None
+
+
+def mpi_initializer(models, logger_level):
+    global experiment_runner
+    experiment_runner = ExperimentRunner(models)
+
+    # Configure logger based on the passed level and adjusted format
+    logging.basicConfig(level=logger_level, format="[%(processName)s/%(levelname)s] %(message)s")
+
+
+class MPIEvaluator(BaseEvaluator):
+    """Evaluator for experiments using MPI Pool Executor from mpi4py"""
+
+    def __init__(self, msis, **kwargs):
+        super().__init__(msis, **kwargs)
+        self._pool = None
+
+    def initialize(self):
+        # Only import mpi4py if the MPIEvaluator is used, to avoid unnecessary dependencies.
+        from mpi4py.futures import MPIPoolExecutor
+
+        # Instead of instantiating the ExperimentRunner for each experiment, instantiate it once here
+        models = NamedObjectMap(AbstractModel)
+        models.extend(self._msis)
+
+        # Use the initializer function to set up the ExperimentRunner for all the worker processes
+        self._pool = MPIPoolExecutor(initializer=mpi_initializer, initargs=(models, _logger.level))
+        _logger.info(f"MPI pool started with {self._pool._max_workers} workers")
+        if self._pool._max_workers <= 10:
+            _logger.warning(
+                f"With only a few workers ({self._pool._max_workers}), the MPIEvaluator may be slower than the Sequential- or MultiprocessingEvaluator"
+            )
+        return self
+
+    def finalize(self):
+        self._pool.shutdown()
+        _logger.info("MPI pool has been shut down")
+
+    def evaluate_experiments(self, scenarios, policies, callback, combine="factorial"):
+        ex_gen = experiment_generator(scenarios, self._msis, policies, combine=combine)
+        experiments = list(ex_gen)  # Convert generator to list
+
+        # Instead of sending all models for each experiment, send only the model_name
+        packed = [(experiment, experiment.model_name) for experiment in experiments]
+
+        # Use the pool to execute in parallel
+        _logger.info(
+            f"MPIEvaluator: Starting {len(packed)} experiments using MPI pool with {self._pool._max_workers} workers"
+        )
+        results = self._pool.map(run_experiment_mpi, packed)
+
+        _logger.info(f"MPIEvaluator: Completed all {len(packed)} experiments")
+        for experiment, outcomes in results:
+            callback(experiment, outcomes)
+        _logger.info(f"MPIEvaluator: Callback completed for all {len(packed)} experiments")
+
+
+def run_experiment_mpi(packed_data):
+    from mpi4py.MPI import COMM_WORLD
+
+    rank = COMM_WORLD.Get_rank()
+
+    experiment, model_name = packed_data
+    _logger.info(f"MPI Rank {rank}: starting {repr(experiment)}")
+
+    # Use the global ExperimentRunner created by the initializer
+    outcomes = experiment_runner.run_experiment(experiment)
+
+    _logger.info(f"MPI Rank {rank}: completed {experiment}")
```

Comment:
same here: why is this an info level log message?

Reply:
Agreed, implemented in 60f1f9c.